dali/c_api_2/data_objects.h

```diff
@@ -0,0 +1,724 @@
+
+    if (!layout) {
+      if (ndim == tl_->sample_dim())
+        new_layout = tl_->GetLayout();
+    } else {
+      new_layout = layout;
+      Validate(new_layout, ndim);
+    }
+
+    tl_->Reset();
+    tl_->SetSize(num_samples);
+    tl_->set_sample_dim(ndim);
+    tl_->SetLayout(new_layout);
+    tl_->set_type(dtype);
+    ptrdiff_t next_offset = 0;
+    auto type_info = TypeTable::GetTypeInfo(dtype);
+    auto element_size = type_info.size();
+
+    std::shared_ptr<void> buffer;
+    if (!deleter.delete_buffer && !deleter.destroy_context) {
+      buffer = std::shared_ptr<void>(data, [](void *){});
+    } else {
+      buffer = std::shared_ptr<void>(data, BufferDeleter{deleter, tl_->order()});
+    }
+
+    bool is_contiguous = true;
+    if (sample_offsets) {
+      for (int i = 0; i < num_samples; i++) {
+        if (sample_offsets[i] != next_offset) {
+          is_contiguous = false;
+          break;
+        }
+        auto num_elements = volume(make_cspan(&shapes[i*ndim], ndim));
+        next_offset += num_elements * element_size;
+      }
+    }
+
+    if (is_contiguous) {
+      tl_->SetContiguity(BatchContiguity::Contiguous);
+      std::vector<int64_t> shape_data(shapes, shapes + ndim * num_samples);
+      TensorListShape<> tl_shape(shape_data, num_samples, ndim);
+      tl_->ShareData(
+        std::move(buffer),
+        next_offset,
+        tl_->is_pinned(),
+        tl_shape,
+        dtype,
+        tl_->device_id(),
+        tl_->order(),
+        new_layout);
+    } else {
+      tl_->SetContiguity(BatchContiguity::Automatic);
+      next_offset = 0;
+
+      for (int i = 0; i < num_samples; i++) {
+        TensorShape<> sample_shape(make_cspan(&shapes[i*ndim], ndim));
+        void *sample_data;
+        size_t sample_bytes = volume(sample_shape) * element_size;
+        if (sample_offsets) {
+          sample_data = static_cast<char *>(data) + sample_offsets[i];
+        } else {
+          sample_data = static_cast<char *>(data) + next_offset;
+          next_offset += sample_bytes;
+        }
+        tl_->SetSample(
+          i,
+          std::shared_ptr<void>(buffer, sample_data),
+          sample_bytes,
+          tl_->is_pinned(),
+          sample_shape,
+          dtype,
+          tl_->device_id(),
+          tl_->order(),
+          new_layout);
+      }
+    }
+  }
+
+  void AttachSamples(
+      int num_samples,
+      int ndim,
+      daliDataType_t dtype,
+      const char *layout,
+      const daliTensorDesc_t *samples,
+      const daliDeleter_t *sample_deleters) override {
+    ValidateNumSamples(num_samples);
+    if (num_samples > 0 && !samples)
+      throw std::invalid_argument("The pointer to sample descriptors must not be NULL.");
+    if (ndim < 0) {
+      if (num_samples == 0)
+        throw std::invalid_argument(
+          "The number of dimensions must not be negative when num_samples is 0.");
+      else
+        ndim = samples[0].ndim;
+      ValidateNDim(ndim);
+    }
```

Comment:
```suggestion
    if (ndim < 0) {
      if (num_samples == 0)
        throw std::invalid_argument(
          "The number of dimensions must not be negative when num_samples is 0.");
      else
        ndim = samples[0].ndim;
    }
    ValidateNDim(ndim);
```

Shouldn't the explicit ndim be validated too?

Reply:
Oh, ok, the check is that it is positive. :D